arch/x86/core/thread.c

```diff
@@ -188,3 +143,95 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
 	thread_monitor_init(thread);
 #endif
 }
+
+#ifdef CONFIG_X86_USERSPACE
+void _x86_swap_update_page_tables(struct k_thread *incoming,
+				  struct k_thread *outgoing)
+{
+	/* Outgoing thread stack no longer accessible */
+	_x86_mmu_set_flags((void *)outgoing->stack_info.start,
+			   ROUND_UP(outgoing->stack_info.size, MMU_PAGE_SIZE),
+			   MMU_ENTRY_SUPERVISOR, MMU_PTE_US_MASK);
+
+
+	/* Userspace can now access the incoming thread's stack */
+	_x86_mmu_set_flags((void *)incoming->stack_info.start,
+			   ROUND_UP(incoming->stack_info.size, MMU_PAGE_SIZE),
+			   MMU_ENTRY_USER, MMU_PTE_US_MASK);
+
+	/* In case of privilege elevation, use the incoming thread's kernel
+	 * stack, the top of the thread stack is the bottom of the kernel stack
+	 */
+	_main_tss.esp0 = incoming->stack_info.start;
+
+	/* TODO: if either thread defines different memory domains, efficiently
+	 * switch between them
+	 */
+}
+
+
+int _arch_is_user_context(void)
+{
+	int cs;
+
+	/* On x86, read the CS register (which cannot be manually set)
+	 * and check the CPL bits within
+	 */
+	__asm__ volatile ("mov %%cs, %[cs_val]" : [cs_val] "=r" (cs));
+
+	/* Mask all but the lowest 2 bits, which is the CPL information */
+	cs &= 0x3;
```

Comment:
FWIW: you could also just define a single "kernel" code segment and test for that descriptor exactly instead of masking off the ring level.  This version is more correct, but takes a few more instructions and is likely to be on a hot path.

Related point: seems like this really wants to be an inline function.  The calling overhead is significantly bigger than the time to execute the meat.

Reply:
good idea on both counts, will do